### PR TITLE
Dynamo DB Kinesis stream adapter option and various fixes

### PR DESCRIPTION
**Pull request by vaslabs**
Hello,

This PR contains changes for the following

- Be able to read dynamo db streams
- A slight change in the way the dynamo db table name is generated since the adapter accepts the full arn which contains characters that are invalid to be in table name
- Consistency in the order of application name and stream name in the configs.
- An example in the readme on how to stream from dynamodb.
- A fix for the initial stream position which seems that it's never passed further and thus ignored.

Cheers,
Vasilis

### CODE REVIEW

File: akka/src/main/scala/com/gilt/gfc/aws/kinesis/akka/KinesisStreamConsumerConfig.scala [hunk 1]
```diff
@@ -1,23 +1,25 @@
...
 import com.amazonaws.services.kinesis.clientlibrary.lib.worker.InitialPositionInStream
 import com.amazonaws.services.kinesis.metrics.impl.NullMetricsFactory
 import com.amazonaws.services.kinesis.metrics.interfaces.IMetricsFactory
 
 import scala.concurrent.duration._
 
 case class KinesisStreamConsumerConfig[T](
-  streamName: String,
   applicationName: String,
+  streamName: String,
```
Comment (gheine) [inline akka/src/main/scala/com/gilt/gfc/aws/kinesis/akka/KinesisStreamConsumerConfig.scala]: Can you revert swapping the argument order? Otherwise existing users of the library who upgrade to the latest version will inadvertently pass in the wrong values.

File: client/src/main/scala/com/gilt/gfc/aws/kinesis/client/KCLConfiguration.scala [hunk 2]
```diff
@@ -36,17 +36,27 @@ object KCLConfiguration {
...
+      kinesisCredentialsProvider,
+      dynamoCredentialsProvider,
+      cloudWatchCredentialsProvider,
+      s"${HostName}:${UUID.randomUUID()}"
     ).withRegionName(regionName.orNull)
+     .withInitialPositionInStream(initialPositionInStream)
+    val adapterConf = dynamoDBKinesisAdapterClient.map {
+      client =>
+        conf.withMaxRecords(1000) //using AWS recommended value
+          .withIdleTimeBetweenReadsInMillis(500) //using AWS recommended value
```
Comment (gheine) [inline client/src/main/scala/com/gilt/gfc/aws/kinesis/client/KCLConfiguration.scala]: Might make sense to make these two configurable too

File: client/src/main/scala/com/gilt/gfc/aws/kinesis/client/KCLWorkerRunner.scala [hunk 3]
```diff
@@ -124,9 +124,18 @@ case class KCLWorkerRunner (
...
-      val worker = metricsFactory.fold(
-        new Worker(recordProcessorFactory, config))(
-        mf => new Worker(recordProcessorFactory, config, mf))
+      val workerBuilder: Worker.Builder = new Worker.Builder()
+            .recordProcessorFactory(recordProcessorFactory)
+              .config(config)
+      dynamoDBKinesisAdapter.foreach(
+        adapter => {
+          workerBuilder.kinesisClient(adapter)
+        }
```
Comment (gheine) [inline client/src/main/scala/com/gilt/gfc/aws/kinesis/client/KCLWorkerRunner.scala]: remove unnecessary curly braces: 
`dynamoDBKinesisAdapter.foreach(adapter => workerBuilder.kinesisClient(adapter))`

File: client/src/main/scala/com/gilt/gfc/aws/kinesis/client/KCLConfiguration.scala [hunk 4]
```diff
@@ -36,17 +36,27 @@ object KCLConfiguration {
...
-    , streamName
-    , kinesisCredentialsProvider
-    , dynamoCredentialsProvider
-    , cloudWatchCredentialsProvider
-    , s"${HostName}:${UUID.randomUUID()}"
+           , dynamoDBKinesisAdapterClient: Option[AmazonDynamoDBStreamsAdapterClient] = None
+           , initialPositionInStream: InitialPositionInStream = InitialPositionInStream.LATEST): KinesisClientLibConfiguration = {
+
+    val dynamoTableName = dynamoDBKinesisAdapterClient.map( _ =>
+    applicationName).getOrElse(s"${applicationName}.${streamName}")
```
Comment (gheine) [inline client/src/main/scala/com/gilt/gfc/aws/kinesis/client/KCLConfiguration.scala]: Could you explain this a bit more? Why the change in table name?

Comment (vaslabs) [inline client/src/main/scala/com/gilt/gfc/aws/kinesis/client/KCLConfiguration.scala]: Hello,
When passing the dynamo stream, the adapter requires the full arn, e.g.
arn:aws:dynamodb:eu-west-1:340721489904:table/cip-masterstack-dqs-EuclidStack-P0GEEOP46WFQ-dynamoDbWallESearchEventTable-QDBYO265Q63S/stream/2017-05-31T12:43:44.127
Which contains characters that are not valid and the dynamo table fails to be created.

So the aim here, is to change the naming strategy if someone passes the adapter.

File: client/src/main/scala/com/gilt/gfc/aws/kinesis/client/KCLConfiguration.scala (hunk 2 above)
Comment (vaslabs) [inline client/src/main/scala/com/gilt/gfc/aws/kinesis/client/KCLConfiguration.scala]: I thought perhaps it would be better to hide this configuration into some sort of advance configs, i.e. "edit this if you really know what you are doing". Do you have any area that this kind of thing can be incorporated

File: akka/src/main/scala/com/gilt/gfc/aws/kinesis/akka/KinesisStreamConsumerConfig.scala (hunk 1 above)
Comment (vaslabs) [inline akka/src/main/scala/com/gilt/gfc/aws/kinesis/akka/KinesisStreamConsumerConfig.scala]: will do

File: client/src/main/scala/com/gilt/gfc/aws/kinesis/client/KCLConfiguration.scala (hunk 2 above)
Comment (gheine) [inline client/src/main/scala/com/gilt/gfc/aws/kinesis/client/KCLConfiguration.scala]: Fair enough. If someone needs to tweak these, they can always submit a PR and make them configurable.

File: client/src/main/scala/com/gilt/gfc/aws/kinesis/client/KCLConfiguration.scala (hunk 4 above)
Comment (gheine) [inline client/src/main/scala/com/gilt/gfc/aws/kinesis/client/KCLConfiguration.scala]: Gotcha. The problem with this is that if the same application uses multiple streams, we'll end up using the same dynamo table for all. 

Alternatively, could we strip all invalid characters from the stream name?

Comment (vaslabs) [inline client/src/main/scala/com/gilt/gfc/aws/kinesis/client/KCLConfiguration.scala]: I think the application name needs a bit of reconsideration. It should be an identifier that the programmer passes, for example in our case we open two kinesis streams from the same application, one to a kinesis stream and one on the dynamo one. For the dynamo one I consider the application name to be the dynamo table name, but that's because I've looked inside to see how it works. 
For now we could strip the characters but for the future I'd have StreamName, ApplicationName and DynamoTableName as value classes to allow a bit of flexibility and some type safety around how the arguments are passed, so that if the order changes for instance it will cause a compile error instead of a hard to debug problem...

To conclude, for now, would you like to replace all ':' and '/' with '-' or just remove them?

**Comment by gheine** [comment]
Released in v0.12.0

File: client/src/main/scala/com/gilt/gfc/aws/kinesis/client/KCLConfiguration.scala (hunk 2 above)
Comment (gheine) [reply client/src/main/scala/com/gilt/gfc/aws/kinesis/client/KCLConfiguration.scala]: Circling back on this, @vaslabs I just realized that the dynamoDBKinesisAdapterClient is never used and by the looks of it can't be passed into KinesisClientLibConfiguration either. Would it make sense to replace the `dynamoDBKinesisAdapterClient: Option[AmazonDynamoDBStreamsAdapterClient] = None` param with two params for maxRecords and idleTimeBetweenReads?

Comment (vaslabs) [reply client/src/main/scala/com/gilt/gfc/aws/kinesis/client/KCLConfiguration.scala]: If I recall right, the only reason is forwarded here is to say, "right, I have to work with the dynamo adapter client, so I just need to set some different configurations." . We could find a better way for modelling this perhaps with two different functions and do the map on a higher level

Comment (vaslabs) [reply client/src/main/scala/com/gilt/gfc/aws/kinesis/client/KCLConfiguration.scala]: I did a PR to resolve this confusion
https://github.com/gilt/gfc-aws-kinesis/pull/37
